util/linuxfw/iptables.go

```diff
@@ -0,0 +1,821 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build linux && !(386 || loong64)
+
+package linuxfw
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"net"
+	"net/netip"
+	"strings"
+	"unsafe"
+
+	"github.com/josharian/native"
+	"golang.org/x/sys/unix"
+	linux_abi "gvisor.dev/gvisor/pkg/abi/linux"
+	"tailscale.com/net/netaddr"
+	"tailscale.com/types/logger"
+)
+
+type sockLen uint32
+
+var (
+	iptablesChainNames = map[int]string{
+		linux_abi.NF_INET_PRE_ROUTING:  "PREROUTING",
+		linux_abi.NF_INET_LOCAL_IN:     "INPUT",
+		linux_abi.NF_INET_FORWARD:      "FORWARD",
+		linux_abi.NF_INET_LOCAL_OUT:    "OUTPUT",
+		linux_abi.NF_INET_POST_ROUTING: "POSTROUTING",
+	}
+	iptablesStandardChains = (func() map[string]bool {
+		ret := make(map[string]bool)
+		for _, v := range iptablesChainNames {
+			ret[v] = true
+		}
+		return ret
+	})()
+)
+
+// DebugNetfilter prints debug information about iptables rules to the
+// provided log function.
+func DebugIptables(logf logger.Logf) error {
+	for _, table := range []string{"filter", "nat", "raw"} {
+		type chainAndEntry struct {
+			chain string
+			entry *entry
+		}
+
+		// Collect all entries first so we can resolve jumps
+		var (
+			lastChain    string
+			ces          []chainAndEntry
+			chainOffsets = make(map[int]string)
+		)
+		err := enumerateIptablesTable(logf, table, func(chain string, entry *entry) error {
+			if chain != lastChain {
+				chainOffsets[entry.Offset] = chain
+				lastChain = chain
+			}
+
+			ces = append(ces, chainAndEntry{
+				chain: lastChain,
+				entry: entry,
+			})
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+
+		lastChain = ""
+		for _, ce := range ces {
+			if ce.chain != lastChain {
+				logf("iptables: table=%s chain=%s", table, ce.chain)
+				lastChain = ce.chain
+			}
+
+			// Fixup jump
+			if std, ok := ce.entry.Target.Data.(standardTarget); ok {
+				if strings.HasPrefix(std.Verdict, "JUMP(") {
+					var off int
+					if _, err := fmt.Sscanf(std.Verdict, "JUMP(%d)", &off); err == nil {
+						if jt, ok := chainOffsets[off]; ok {
+							std.Verdict = "JUMP(" + jt + ")"
+							ce.entry.Target.Data = std
+						}
+					}
+				}
+			}
+
+			logf("iptables:   entry=%+v", ce.entry)
+		}
+	}
+	return nil
+}
+
+// DetectIptables returns the number of iptables rules that are present in the
+// system, ignoring the default "ACCEPT" rule present in the standard iptables
+// chains.
+func DetectIptables() (int, error) {
```

Comment:
document: when does it return an error?

on 0? on error? on not supported by kernel?